content/en/docs/use-cases/encrypted-images.md

```diff
@@ -0,0 +1,131 @@
+---
+title: Encrypted images
+date: 2023-01-24
+description: Procedures to encrypt and consume OCI images in a TEE
+categories:
+- use case
+tags:
+- coco-keyprovider
+- images
+- kbs
+---
+
+# Using encrypted images in CoCo
+
+A user might want to bundle sensitive data on an OCI (Docker) image. The image layers should only be accessible within a Trusted Execution Environment (TEE).
+
+The project provides the means to encrypt an image with a symmetric key that is released to the TEE only after successful verification and appraisal in a Remote Attestation process. CoCo infrastructure components within the TEE will transparently decrypt the image layers as they are pulled from a registry without exposing them to the outside.
+
+## Instructions
+
+The following steps require a functional CoCo installation on a Kubernetes cluster. A Key Broker Client (KBC) has to be configured for TEEs to be able to retrieve confidential secrets. We assume `cc_kbc` as a KBC for the CoCo Projects KBS in the following instructions, but image encryption should work with other Key Broker implementations similarly.
+
+Please make sure you have a recent version of (skopeo)[https://github.com/containers/skopeo] (v1.14+) installed locally.
+
+### Encrypt an image
+
+We extend public image with secret data.
+
+```bash
+docker build -t unencrypted - <<EOF
+FROM nginx:stable
+RUN echo "something confidential" > /secret
+EOF
+```
+
+The encryption key needs to be a 32 byte sequence and provided to the encryption step as base64-encoded string.
+
+The key id is a generic resource descriptor used by the Key Broker to look up secrets in its storage. For CoCo's KBS this is composed from three segments: `$repository_name/$resource_type/$resource_tag`
+
+```bash
+head -c 32 /dev/urandom | openssl enc > image_key
+KEY_B64="$(base64 < image_key)"
+KEY_PATH="/default/image_key/nginx"
+KEY_ID="kbs://${KEY_PATH}"
+```
+
+The image encryption logic is invoked in a container. To access the unencrypted image from within the container, skopeo can be used to buffer the image in a directory, which is then made available to the container. Similary, the resulting encrypted image will be put into an output directory.
+
+```bash
+mkdir -p oci/{input,output}
+skopeo copy docker-daemon:unencrypted:latest dir:./oci/input
```

Comment:
```bash
docker --version 
Docker version 24.0.7, build afdd53b
```

odd, what's your skopeo version? if skopeo is not able to access local images with docker v25, we need to adjust the steps (but I suppose skopeo wants to support this?)

Reply:
```
$ skopeo --version
skopeo version 1.14.1 commit: 45b7bf5e4ac70a5bfe29e0486719d8390bd6679c
```

Reply:
that's the latest version, so it doesn't work with the latest docker, that's a bummer :( Or is that maybe due to docker on mac weirdness?

Reply:
No this is on a linux on x86 machine. No docker on mac.

Reply:
it's a skopeo issue. there's a workaround, but a fixed skopeo release is probably pending. we can wait for a couple of days maybe.

https://github.com/containers/skopeo/issues/2202

Reply:
I was able to downgrade docker using `sudo apt install docker-ce=5:24.0.8-1~ubuntu.22.04~jammy` but now facing another issue: 

```console
$ docker run -v "${PWD}/oci:/oci"     ghcr.io/confidential-containers/coco-keyprovider        /encrypt.sh -k "$KEY_B64" -i "$KEY_ID" -s dir:/oci/input -d dir:/oci/output
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/encrypt.sh": stat /encrypt.sh: no such file or directory: unknown.
ERRO[0000] error waiting for container: context canceled
```

```console
$ docker images | grep coco-keyprovider
ghcr.io/confidential-containers/coco-keyprovider    latest    cb48099857de   7 months ago   90.8MB
```

And the image has no `/encrypt.sh` file:

```console
$ docker run -it ghcr.io/confidential-containers/coco-keyprovider ls /
bin  boot  dev  etc  home  lib  lib32  lib64  libx32  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
```

Reply:
ouch, yes. 🙈  this is dependent on a pending PR in guest-components: https://github.com/confidential-containers/guest-components/pull/451

we can park this PR until this gets merged, but if you want to test it anyway:

```bash
git clone https://github.com/confidential-containers/guest-components.git
cd guest-components
gh pr checkout 451
docker build -t  ghcr.io/confidential-containers/coco-keyprovider -f attestation-agent/docker/Dockerfile.keyprovider .
```

Reply:
another option would be to not use an encrypt.sh convenience script in the reference, but spell out all the skopeo/coco-steps here. In my experience this hasn't really worked well, since it's very easy to make a mistake and you end up with a non-working image that is hard to debug.

Reply:
Ah, I misread the workflow. the PR is merged but an official coco-keyprovider image will only be built once a new release has been cut. So you have to build the image in any case (until the next release) we can maybe make the docker build part of the instructions?

Reply:
there will be no keyprovider release soon-ish, and the ghcr.io repo has been archived, so I added the build instructions here.